<a href="https://colab.research.google.com/github/vjmiyagi/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/module3-join-and-reshape-data/LS_DS14_113_Join_and_Reshape_Data_Assignment.ipynbLS_DS14_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [1]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-03-11 18:56:39--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.146.93
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.146.93|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  43.4MB/s    in 4.4s    

2020-03-11 18:56:44 (44.7 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [2]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [3]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [7]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [31]:
##### YOUR CODE HERE #####

import pandas as pd

# Check out the shape of the CSV files

aisles = pd.read_csv('aisles.csv')
print(aisles.shape)
aisles.head()
# aisles.csv is not needed


(134, 2)


aisle_id                       aisle
0         1       prepared soups salads
1         2           specialty cheeses
2         3         energy granola bars
3         4               instant foods
4         5  marinades meat preparation

In [32]:
departments = pd.read_csv('departments.csv')
print(departments.shape)
departments.head()

# departments.csv is not needed

(21, 2)


department_id department
0              1     frozen
1              2      other
2              3     bakery
3              4    produce
4              5    alcohol

In [33]:
order_products_prior = pd.read_csv('order_products__prior.csv')
print(order_products_prior.shape)
order_products_prior.head()

# product_id is needed from order_products_prior

(32434489, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [34]:
order_products_train = pd.read_csv('order_products__train.csv')
print(order_products_train.shape)
order_products_train.head()

# product_id is needed

(1384617, 4)


order_id  product_id  add_to_cart_order  reordered
0         1       49302                  1          1
1         1       11109                  2          1
2         1       10246                  3          0
3         1       49683                  4          0
4         1       43633                  5          1

In [35]:
orders = pd.read_csv('orders.csv')
print(orders.shape)
orders.head()

(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [36]:
products = pd.read_csv('products.csv')
print(products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

1.   Banana
2.   Bag of Organic Bananas
3.   Organic Strawberries
4.   Organic Baby Spinach
5.   Organic Hass Avocado
6.   Organic Avocado
7.   Large Lemon
8.   Strawberries
9.   Limes
10.  Organic Whole Milk

In [37]:
condition1 = (products['product_name'] == "Banana")
condition2 = (products['product_name'] == "Bag of Organic Bananas")
condition3 = (products['product_name'] == "Organic Strawberries")
condition4 = (products['product_name'] == "Organic Baby Spinach")
condition5 = (products['product_name'] == "Organic Hass Avocado")
condition6 = (products['product_name'] == "Organic Avocado")
condition7 = (products['product_name'] == "Large Lemon")
condition8 = (products['product_name'] == "Strawberries")
condition9 = (products['product_name'] == "Limes")
condition10 = (products['product_name'] == "Organic Whole Milk")

products_subset = products[condition1 | condition2 |
                           condition3 | condition4 |
                           condition5 | condition6 |
                           condition7 | condition8 |
                           condition9 | condition10]

products_subset.head(10)

product_id            product_name  aisle_id  department_id
13175       13176  Bag of Organic Bananas        24              4
16796       16797            Strawberries        24              4
21136       21137    Organic Strawberries        24              4
21902       21903    Organic Baby Spinach       123              4
24851       24852                  Banana        24              4
26208       26209                   Limes        24              4
27844       27845      Organic Whole Milk        84             16
47208       47209    Organic Hass Avocado        24              4
47625       47626             Large Lemon        24              4
47765       47766         Organic Avocado        24              4

In [38]:
# Assign product_id to variables for top ten items

prod1 = 24852
prod2 = 13176
prod3 = 21137
prod4 = 21903
prod5 = 47209
prod6 = 47766
prod7 = 47626
prod8 = 16797
prod9 = 26209
prod10 = 27845

# Set conditions for order_products_prior

p1_cond1 = (order_products_prior['product_id'] == prod1)
p1_cond2 = (order_products_prior['product_id'] == prod2)
p1_cond3 = (order_products_prior['product_id'] == prod3)
p1_cond4 = (order_products_prior['product_id'] == prod4)
p1_cond5 = (order_products_prior['product_id'] == prod5)
p1_cond6 = (order_products_prior['product_id'] == prod6)
p1_cond7 = (order_products_prior['product_id'] == prod7)
p1_cond8 = (order_products_prior['product_id'] == prod8)
p1_cond9 = (order_products_prior['product_id'] == prod9)
p1_cond10 = (order_products_prior['product_id'] == prod10)




order_products_prior_subset = order_products_prior[p1_cond1 |
                                                    p1_cond2 |
                                                    p1_cond3 |
                                                    p1_cond4 |
                                                    p1_cond5 |
                                                    p1_cond6 |
                                                    p1_cond7 |
                                                    p1_cond8 |
                                                    p1_cond9 |
                                                    p1_cond10]
print(order_products_prior_subset.shape)
order_products_prior_subset.head()

(2323158, 4)


order_id  product_id  add_to_cart_order  reordered
12         3       21903                  4          1
30         5       13176                  1          1
51         5       47209                 22          0
77        10       24852                  1          1
80        10       47766                  4          1

In [39]:
# Set conditions for order_products_train

p2_cond1 = (order_products_train['product_id'] == prod1)
p2_cond2 = (order_products_train['product_id'] == prod2)
p2_cond3 = (order_products_train['product_id'] == prod3)
p2_cond4 = (order_products_train['product_id'] == prod4)
p2_cond5 = (order_products_train['product_id'] == prod5)
p2_cond6 = (order_products_train['product_id'] == prod6)
p2_cond7 = (order_products_train['product_id'] == prod7)
p2_cond8 = (order_products_train['product_id'] == prod8)
p2_cond9 = (order_products_train['product_id'] == prod9)
p2_cond10 = (order_products_train['product_id'] == prod10)




order_products_train_subset = order_products_train[p2_cond1 |
                                                   p2_cond2 |
                                                   p2_cond3 |
                                                   p2_cond4 |
                                                   p2_cond5 |
                                                   p2_cond6 |
                                                   p2_cond7 |
                                                   p2_cond8 |
                                                   p2_cond9 |
                                                   p2_cond10]
print(order_products_train_subset.shape)
order_products_train_subset.head()

(95156, 4)


order_id  product_id  add_to_cart_order  reordered
5          1       13176                  6          0
6          1       47209                  7          0
35        98       13176                  4          1
85       112       47209                  5          0
94       170       13176                  3          1

In [43]:
# Start putting these dataframes together

df1 = order_products_prior_subset
print(df1.head())

df2 = order_products_train_subset
print(df2.head())

order_products = pd.concat([df1, df2])

print(order_products.shape)
order_products.head()

    order_id  product_id  add_to_cart_order  reordered
12         3       21903                  4          1
30         5       13176                  1          1
51         5       47209                 22          0
77        10       24852                  1          1
80        10       47766                  4          1
    order_id  product_id  add_to_cart_order  reordered
5          1       13176                  6          0
6          1       47209                  7          0
35        98       13176                  4          1
85       112       47209                  5          0
94       170       13176                  3          1
(2418314, 4)


order_id  product_id  add_to_cart_order  reordered
12         3       21903                  4          1
30         5       13176                  1          1
51         5       47209                 22          0
77        10       24852                  1          1
80        10       47766                  4          1

In [0]:
# Add products names to make more sense of it all

final = pd.merge(order_products, products_subset, on='product_id', how='inner')

In [50]:
final


order_id  product_id  ...  aisle_id  department_id
0               3       21903  ...       123              4
1              26       21903  ...       123              4
2              31       21903  ...       123              4
3              39       21903  ...       123              4
4              56       21903  ...       123              4
...           ...         ...  ...       ...            ...
2418309   3418861       26209  ...        24              4
2418310   3418905       26209  ...        24              4
2418311   3419642       26209  ...        24              4
2418312   3420257       26209  ...        24              4
2418313   3420286       26209  ...        24              4

[2418314 rows x 7 columns]

In [52]:
# df.drop(['B', 'C'], axis=1)

final = final.drop(['order_id', 'add_to_cart_order', 'reordered', 'aisle_id', 
                    'department_id'], axis=1)

final

product_id          product_name
0             21903  Organic Baby Spinach
1             21903  Organic Baby Spinach
2             21903  Organic Baby Spinach
3             21903  Organic Baby Spinach
4             21903  Organic Baby Spinach
...             ...                   ...
2418309       26209                 Limes
2418310       26209                 Limes
2418311       26209                 Limes
2418312       26209                 Limes
2418313       26209                 Limes

[2418314 rows x 2 columns]

In [57]:
# Please pardon my less fluffy variable name
stuff = final['product_name']
stuff.value_counts()

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [58]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
# table2 -> tidy

# Take the row Index and add it as a new column

table2 = table2.reset_index()

In [60]:
table2

index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [63]:
tidy2 = table2.melt(id_vars='index', value_vars=['John Smith', 'Jane Doe', 'Mary Johnson'])

tidy2

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [65]:
# Rename the columns

tidy2 = tidy2.rename(columns={
    'index': 'trt',
    'variable': 'name',
    'value': 'result'
})
tidy2

trt          name  result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [66]:
# Reorder columns

tidy2 = tidy2[['name', 'trt', 'result']]

tidy2

name         trt  result
0    John Smith  treatmenta     NaN
1    John Smith  treatmentb     2.0
2      Jane Doe  treatmenta    16.0
3      Jane Doe  treatmentb    11.0
4  Mary Johnson  treatmenta     3.0
5  Mary Johnson  treatmentb     1.0

In [67]:
# Clean up trt column

tidy2.trt = tidy2.trt.str.replace('treatment', '')

tidy2

name trt  result
0    John Smith   a     NaN
1    John Smith   b     2.0
2      Jane Doe   a    16.0
3      Jane Doe   b    11.0
4  Mary Johnson   a     3.0
5  Mary Johnson   b     1.0

In [76]:
# I would make it even tidier by replacing the NaN with a 0 and make int

tidy2.fillna(0, inplace=True)

tidy2

name trt  result
0    John Smith   a     0.0
1    John Smith   b     2.0
2      Jane Doe   a    16.0
3      Jane Doe   b    11.0
4  Mary Johnson   a     3.0
5  Mary Johnson   b     1.0

In [78]:
# Change to integer

tidy2.result = tidy2.result.astype(int)

tidy2

name trt  result
0    John Smith   a       0
1    John Smith   b       2
2      Jane Doe   a      16
3      Jane Doe   b      11
4  Mary Johnson   a       3
5  Mary Johnson   b       1

In [80]:
# tidy -> table2

wide = tidy2.pivot_table(index='name', columns='trt', values='result')

wide

trt            a   b
name                
Jane Doe      16  11
John Smith     0   2
Mary Johnson   3   1

In [0]:
import seaborn as sns

In [84]:
flights = sns.load_dataset('flights')

flights

year      month  passengers
0    1949    January         112
1    1949   February         118
2    1949      March         132
3    1949      April         129
4    1949        May         121
..    ...        ...         ...
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432

[144 rows x 3 columns]

In [0]:
# Flights Pivot Table

Load seaborn's flights dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.


In [85]:
wide_f = flights.pivot_table(index='year', columns='month', values='passengers')

wide_f

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####